### PR TITLE
JVM: refactor LambdaInfo a bit more

### PR DESCRIPTION
**Pull request by pyos**
* Inliner will now ensure consistency of `invokeMethodParameters` instead of silently generating incorrect code for inline class parameters.
* `DefaultLambda` no longer uses descriptors (`PsiDefaultLambda` does)
* `IrDefaultLambda.invokeMethodParameters` will now correctly return non-erased types for function references and lambdas, which means the inliner will unbox inline classes when inlining them.
* The code is just shorter in general (why use a `ClassVisitor` when `ClassReader.superName` is enough?)
* If any parameters of default lambdas are inline classes, the inliner will try to mangle then name of the invoke method before falling back to the unmangled one. Old compiler versions still won't be able to load default lambdas generated by JVM_IR, but this way we avoid incorrect behavior of function references taking inline class types that unbox to Any at least.

#KT-46601 Fixed

### CODE REVIEW

**Comment by max-kammerer** [comment]
FirBlackBoxInlineCodegenTestGenerated$DefaultValues$LambdaInlining$CallableReferences.testWithInlineClassParameter test failed:

```
  Caused by: org.jetbrains.kotlin.codegen.CompilationException: Back-end (JVM) Internal error: Couldn't inline method call 'inlineFun-1BDWgbU$default' into
  public fun box(): kotlin.String defined in root package
  <no source>
  Cause: inlineFun-1BDWgbU$default (Ljava/lang/String;Lkotlin/jvm/functions/Function1;ILjava/lang/Object;)Ljava/lang/String;:
  L0
  L1
  LINENUMBER 12 L1
  INVOKESTATIC $$$DEFAULT_LAMBDA_FAKE_CALL$$$.$$$DEFAULT_LAMBDA_FAKE_CALL$$$1 ()V
  L2
  L3
  L4
  ALOAD 0
  LDC "c"
  INVOKESTATIC kotlin/jvm/internal/Intrinsics.checkNotNullParameter (Ljava/lang/Object;Ljava/lang/String;)V
  ALOAD 1
  LDC "x"
  INVOKESTATIC kotlin/jvm/internal/Intrinsics.checkNotNullParameter (Ljava/lang/Object;Ljava/lang/String;)V
  ICONST_0
  ISTORE 2
  L5
  ALOAD 1
  L6
  LINENUMBER 12 L6
  ALOAD 0
  INVOKESTATIC test/C.box-impl (Ljava/lang/String;)Ltest/C;
  INVOKEINTERFACE kotlin/jvm/functions/Function1.invoke (Ljava/lang/Object;)Ljava/lang/Object; (itf)
  CHECKCAST java/lang/String
  ARETURN
  L7
  NOP
  L8
  ARETURN
  L9
  NOP
  L10
  LOCALVARIABLE $i$f$inlineFun-1BDWgbU I L5 L8 2
  LOCALVARIABLE c Ljava/lang/String; L0 L10 0
  MAXSTACK = 2
  MAXLOCALS = 4
  File is unknown
  The root cause java.lang.IllegalStateException was thrown at: org.jetbrains.kotlin.codegen.inline.DefaultLambda.generateLambdaBody(LambdaInfo.kt:137)
    at org.jetbrains.kotlin.codegen.inline.InlineCodegen.throwCompilationException(InlineCodegen.kt:101)
    at org.jetbrains.kotlin.codegen.inline.InlineCodegen.performInline(InlineCodegen.kt:141)
    at org.jetbrains.kotlin.backend.jvm.codegen.IrInlineCodegen.genInlineCall(IrInlineCodegen.kt:144)
    at org.jetbrains.kotlin.backend.jvm.codegen.IrInlineCallGenerator$DefaultImpls.genCall(IrInlineCallGenerator.kt:29)

....
  Caused by: java.lang.IllegalStateException: Can't find method 'invoke(Ljava/lang/Object;)Ljava/lang/Object;' in 'test/_1Kt$inlineFun$1'
    at org.jetbrains.kotlin.codegen.inline.DefaultLambda.generateLambdaBody(LambdaInfo.kt:137)
    at org.jetbrains.kotlin.codegen.inline.InlineCodegen.generateClosuresBodies(InlineCodegen.kt:351)
    at org.jetbrains.kotlin.codegen.inline.InlineCodegen.inlineCall(InlineCodegen.kt:212)
    at org.jetbrains.kotlin.codegen.inline.InlineCodegen.performInline(InlineCodegen.kt:135)
```

**Comment by pyos** [comment]
Right, it's broken on JVM_IR but I forgot to mute it for FIR. Fixed.

**Comment by pyos** [comment]
On second thought, here's a fix for that issue.

**Comment by pyos** [comment]
@max-kammerer ping? this is needed to fix KT-45505

File: compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt [hunk 1]
```diff
@@ -130,84 +100,52 @@ abstract class DefaultLambda(
...
-        get() = invokeMethodDescriptor.returnType
-
     final override lateinit var capturedVars: List<CapturedParamDesc>
         private set
 
     var originalBoundReceiverType: Type? = null
         private set
 
-    override val isSuspend = parameterDescriptor.isSuspendLambda
+    override val isSuspend = false // TODO: it should probably be true sometimes, but it never was
```
Comment (max-kammerer) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt:109]: FYI: Frontend reports: "Suspend functional parameters with default values are not yet supported in inline functions"

Comment (max-kammerer) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt:109]: Could you mention it here?

File: compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt [hunk 2]
```diff
@@ -130,84 +100,52 @@ abstract class DefaultLambda(
     final override lateinit var invokeMethod: Method
         private set
 
-    private lateinit var invokeMethodDescriptor: FunctionDescriptor
-
-    override val invokeMethodParameters: List<KotlinType?>
-        get() = invokeMethodDescriptor.valueParameters.map { it.returnType } // should be FunctionN, so no extension receiver
```
Comment (max-kammerer) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt:136]: Is old assumption is broken?

Comment (max-kammerer) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt:136]: I mean ` // should be FunctionN, so no extension receiver` ?

File: compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt [hunk 3]
```diff
@@ -130,84 +100,52 @@ abstract class DefaultLambda(
...
-
-                super.visit(version, access, name, signature, superName, interfaces)
-            }
-        }, ClassReader.SKIP_CODE or ClassReader.SKIP_FRAMES or ClassReader.SKIP_DEBUG)
-
-        invokeMethodDescriptor = findInvokeMethodDescriptor(isPropertyReference)
-
-        val descriptor = Type.getMethodDescriptor(Type.VOID_TYPE, *capturedArgs)
-        val constructor = getMethodNode(classBytes, "<init>", descriptor, lambdaClassType)?.node
+        val superName = ClassReader(classBytes).superName
```
Comment (max-kammerer) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt:113]: Cool, I've missed it

File: compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt (hunk 2 above)
Comment (pyos) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt:136]: It's fine, it's just safer this way I think

File: compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt (hunk 1 above)
Comment (pyos) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt:109]: Huh, it really does have this error, but the `functionDescriptor.isSuspend` check means that it's only reported on `suspend inline fun f(x: suspend () -> Unit = {})` and not `inline fun f(crossinline x: suspend () -> Unit = {})` even though both are broken. Using suspend *function references* works though, the problem is specifically with suspend lambdas (there's no version of `invoke` without the state machine, so they're completely uninlinable).

Comment (max-kammerer) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/inline/LambdaInfo.kt:109]: OK, thank you for detailed explanation

**Comment by max-kammerer** [comment]
@pyos Thank you!

**Pull request merged**